idea/idea-fir/src/org/jetbrains/kotlin/idea/core/overrideImplement/KtGenerateMembersHandler.kt

```diff
@@ -0,0 +1,319 @@
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtCallableSymbol
+import org.jetbrains.kotlin.idea.frontend.api.tokens.HackToForceAllowRunningAnalyzeOnEDT
+import org.jetbrains.kotlin.idea.frontend.api.tokens.hackyAllowRunningOnEdt
+import org.jetbrains.kotlin.idea.util.application.runWriteAction
+import org.jetbrains.kotlin.psi.KtCallableDeclaration
+import org.jetbrains.kotlin.psi.KtClassBody
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.psi.KtDeclaration
+import org.jetbrains.kotlin.psi.psiUtil.endOffset
+import org.jetbrains.kotlin.psi.psiUtil.prevSiblingOfSameType
+import org.jetbrains.kotlin.psi.psiUtil.startOffset
+import org.jetbrains.kotlin.utils.addToStdlib.firstIsInstanceOrNull
+import org.jetbrains.kotlin.utils.addToStdlib.firstNotNullResult
+
+internal abstract class KtGenerateMembersHandler : AbstractGenerateMembersHandler<KtClassMember>() {
+    @OptIn(HackToForceAllowRunningAnalyzeOnEDT::class)
+    override fun generateMembers(
+        editor: Editor,
+        classOrObject: KtClassOrObject,
+        selectedElements: Collection<KtClassMember>,
+        copyDoc: Boolean
+    ) {
+        // Using hackyAllowRunningOnEdt here because we don't want to pre-populate all possible textual overrides before user selection.
+        val (commands, insertedBlocks) = hackyAllowRunningOnEdt {
+            val insertedBlocks = analyse(classOrObject) {
+                this.generateMembers(editor, classOrObject, selectedElements, copyDoc)
+            }
+            // Reference shortening is done in a separate analysis session because the session need to be aware of the newly generated
+            // members.
+            val commands = analyse(classOrObject) {
+                insertedBlocks.mapNotNull { block ->
+                    val declarations = block.declarations.mapNotNull { it.element }
+                    val first = declarations.firstOrNull() ?: return@mapNotNull null
+                    val last = declarations.last()
+                    collectPossibleReferenceShortenings(first.containingKtFile, TextRange(first.startOffset, last.endOffset))
+                }
+            }
+            commands to insertedBlocks
+        }
+        runWriteAction {
+            commands.forEach { it.invokeShortening() }
+            val project = classOrObject.project
+            val codeStyleManager = CodeStyleManager.getInstance(project)
+            insertedBlocks.forEach { it.declarations.forEach { it.element?.let { element -> codeStyleManager.reformat(element) } } }
+            insertedBlocks.firstOrNull()?.declarations?.firstNotNullResult { it.element }?.let {
+                moveCaretIntoGeneratedElement(editor, it)
+            }
+        }
+    }
+
+    /**
+     * @returns the generated PSI element inside the target class. The members are grouped together if they are near each other. Each group contains the members in the order they appear in the target class after code generation
```

Comment:
Nit: this line looks too long?

Reply:
OK removed since the doc on `InsertedBlock` should be informative enough.